src/audio_pixel.cpp

```diff
@@ -0,0 +1,276 @@
+#include "audio_pixel.h"
+#define project nullptr
+
+// audio pixel functions 
+
+double linear_interp(double x, double x1, double x2, double y1, double y2) {
+    // edge cases 
+    if ((x2-x1) == 0)
+        return x1;
+    // interpolation in time domain
+    return (((x2 - x) / (x2 - x1)) * y1) + ((x - x1) / (x2 - x1)) * y2;
+}
+
+double closest_val(double val1, double val2, double target) {
+    return abs(target - val1) <= abs(target - val2) ? val1 : val2;
+}
+
+template<typename T>
+const std::vector<T> get_view(std::vector<T> const &parent, int start, int end)
+{
+    // todo: make sure start and  end indices are valid
+    auto startptr = v.cbegin() + start;
+    auto endptr = v.cbegin() + end;
+ 
+    std::vector<T> vec(startptr, endptr);
+    return vec;
+}
+
+// ************* audio_pixel_block_t ****************
+
+bool audio_pixel_block_t::flush() const {
+    json j;
+    to_json(j);
+    std::string resource_path = GetResourcePath();
+    std::ofstream ofs;
+    ofs.open(resource_path + "/kiwi-block.json");
+
+    if (!ofs.is_open()){
+        return false;
+    }
+    
+    ofs << std::setw(4) << j << std::endl;
+    return true;
+}
+
+void audio_pixel_block_t::to_json(json& j) const {
+    j[std::to_string(m_pix_per_s)] = m_channel_pixels;
+}
+
+void audio_pixel_block_t::update(std::vector<double>& sample_buffer, int num_channels, int sample_rate) {
+    // constructs an audio pixel block using interleaved sample buffer and num channels
+    // clear the current set of audio pixels 
+    m_channel_pixels.clear();
+    int pixels_per_channel = 0;
+
+    // initalize the m_channel_pixels to have empty audio pixel vectors
+    for (int i = 0; i < num_channels; i++) {
+        std::vector<audio_pixel_t> empty_pixel_vec;
+        m_channel_pixels.push_back(empty_pixel_vec);
+    }
+
+    //calcualte the samples per audio pixel and initalize an audio pixel
+    int samples_per_pixel = sample_rate / m_pix_per_s;
+    m_block_size = ceil((sample_buffer.size()/num_channels)/m_pix_per_s);
```

Comment:
shouldn't the block size be `ceil(sample_buffer.size() / num_channels) / samples_per_pixel`? 
hmm

Reply:
also, why do we need to store  the block size as a member? isn't that information already in `m_channel_pixels`? 

```cpp
int block_size = m_channel_pixels.empty() ? 0 : m_channel_pixels.front().size() ; 
```

Reply:
true thank you for catching this

Reply:
👑